File: import-to-hubspot.js
import https from 'node:https';
import { loadEnv, sleep, readJSON, writeJSON } from './lib.js';

const env = loadEnv();
const HUBSPOT_TOKEN = env.HUBSPOT_API_TOKEN;
const ENRICHED_FILE = 'data/enriched.json';
const PROGRESS_FILE = 'data/.import-progress';

const BATCH_SIZE = 10;
const BATCH_DELAY = 1000;

function batchCreate(contacts) {
  const inputs = contacts.map((c) => ({
    properties: {
      email: c.email,
      firstname: c.first_name,
      lastname: c.last_name,
      jobtitle: c.jobtitle || '',
      company: c.company || '',
      hs_linkedin_url: c.linkedin_url || '',
      do_not_email: 'true',
    },
  }));

  return new Promise((resolve, reject) => {
    const payload = JSON.stringify({ inputs });
    const opts = {
      hostname: 'api.hubapi.com',
      path: '/crm/v3/objects/contacts/batch/create',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${HUBSPOT_TOKEN}`,
      },
    };
    const req = https.request(opts, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        if (res.statusCode === 429) {
          reject(new Error('RATE_LIMITED'));
          return;
        }
        try {
          const parsed = JSON.parse(data);
          resolve({ status: res.statusCode, body: parsed });
        } catch {
          resolve({ status: res.statusCode, body: data });
        }
      });
    });
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

async function main() {
  const enriched = readJSON(ENRICHED_FILE);
  if (enriched.length === 0) {
    console.error('No enriched contacts found. Run enrich-contacts.js first.');
    process.exit(1);
  }

  // Resume support: load last completed batch index
  let startIndex = 0;
  try {
    const saved = readJSON(PROGRESS_FILE);
    if (typeof saved === 'number') startIndex = saved;
  } catch {}

  const totalBatches = Math.ceil(enriched.length / BATCH_SIZE);
  const startBatch = Math.floor(startIndex / BATCH_SIZE);
  console.log(`Importing ${enriched.length} contacts in ${totalBatches} batches of ${BATCH_SIZE}`);
  if (startBatch > 0) console.log(`Resuming from batch ${startBatch + 1}`);

  let created = 0;
  let duplicates = 0;
  let errors = 0;

  for (let batch = startBatch; batch < totalBatches; batch++) {
    const start = batch * BATCH_SIZE;
    const end = Math.min(start + BATCH_SIZE, enriched.length);
    const slice = enriched.slice(start, end);

    console.log(`Batch ${batch + 1}/${totalBatches} (contacts ${start + 1}-${end})...`);

    let retries = 0;
    while (true) {
      try {
        const res = await batchCreate(slice);

        if (res.status === 201) {
          created += slice.length;
          console.log(`  ✓ Created ${slice.length} contacts`);
        } else if (res.status === 409 || res.status === 207) {
          // Partial success or conflicts
          const results = res.body.results || [];
          const errs = res.body.errors || [];
          created += results.length;
          duplicates += errs.length;
          console.log(`  Partial: ${results.length} created, ${errs.length} duplicates/errors`);
        } else {
          console.log(`  Unexpected status ${res.status}: ${JSON.stringify(res.body).slice(0, 200)}`);
          errors += slice.length;
        }
        break;
      } catch (err) {
        if (err.message === 'RATE_LIMITED') {
          retries++;
          const backoff = Math.min(1000 * Math.pow(2, retries), 60000);
          console.log(`  Rate limited. Backing off ${backoff / 1000}s...`);
          await sleep(backoff);
          continue;
        }
        console.error(`  Error: ${err.message}`);
        errors += slice.length;
        break;
      }
    }

    // Save progress
    writeJSON(PROGRESS_FILE, end);
    await sleep(BATCH_DELAY);
  }

  console.log(`\nDone.`);
  console.log(`  Created:    ${created}`);
  console.log(`  Duplicates: ${duplicates}`);
  console.log(`  Errors:     ${errors}`);
  console.log(`\nNext: In HubSpot UI, filter contacts by contact_source = "linkedin-import" and set them as non-marketing.`);
}

main();

File: import_to_hubspot.py
#!/usr/bin/env python3
"""Batch import enriched contacts into HubSpot CRM."""

import math
import os
import sys

import requests

from lib import init_env, read_json, sleep_ms, write_json

init_env()

HUBSPOT_TOKEN = os.environ.get('HUBSPOT_API_TOKEN', '')
ENRICHED_FILE = 'data/enriched.json'
PROGRESS_FILE = 'data/.import-progress'

BATCH_SIZE = 10
BATCH_DELAY = 1000


def batch_create(contacts):
    inputs = [
        {
            'properties': {
                'email': c['email'],
                'firstname': c['first_name'],
                'lastname': c['last_name'],
                'jobtitle': c.get('jobtitle', ''),
                'company': c.get('company', ''),
                'hs_linkedin_url': c.get('linkedin_url', ''),
                'do_not_email': 'true',
            }
        }
        for c in contacts
    ]

    resp = requests.post(
        'https://api.hubapi.com/crm/v3/objects/contacts/batch/create',
        json={'inputs': inputs},
        headers={
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {HUBSPOT_TOKEN}',
        },
        timeout=30,
    )
    if resp.status_code == 429:
        raise Exception('RATE_LIMITED')
    try:
        body = resp.json()
    except Exception:
        body = resp.text
    return {'status': resp.status_code, 'body': body}


def main():
    enriched = read_json(ENRICHED_FILE)
    if not enriched:
        print('No enriched contacts found. Run enrich_contacts.py first.')
        sys.exit(1)

    # Resume support
    start_index = 0
    try:
        saved = read_json(PROGRESS_FILE)
        if isinstance(saved, int):
            start_index = saved
    except Exception:
        pass

    total_batches = math.ceil(len(enriched) / BATCH_SIZE)
    start_batch = start_index // BATCH_SIZE
    print(f'Importing {len(enriched)} contacts in {total_batches} batches of {BATCH_SIZE}')
    if start_batch > 0:
        print(f'Resuming from batch {start_batch + 1}')

    created = 0
    duplicates = 0
    errors = 0

    for batch_num in range(start_batch, total_batches):
        start = batch_num * BATCH_SIZE
        end = min(start + BATCH_SIZE, len(enriched))
        batch = enriched[start:end]

        print(f'Batch {batch_num + 1}/{total_batches} (contacts {start + 1}-{end})...')

        retries = 0
        while True:
            try:
                res = batch_create(batch)

                if res['status'] == 201:
                    created += len(batch)
                    print(f'  Created {len(batch)} contacts')
                elif res['status'] in (409, 207):
                    results = res['body'].get('results', []) if isinstance(res['body'], dict) else []
                    errs = res['body'].get('errors', []) if isinstance(res['body'], dict) else []
                    created += len(results)
                    duplicates += len(errs)
                    print(f'  Partial: {len(results)} created, {len(errs)} duplicates/errors')
                else:
                    body_str = str(res['body'])[:200]
                    print(f'  Unexpected status {res["status"]}: {body_str}')
                    errors += len(batch)
                break
            except Exception as err:
                if str(err) == 'RATE_LIMITED':
                    retries += 1
                    backoff = min(1000 * (2 ** retries), 60000)
                    print(f'  Rate limited. Backing off {backoff / 1000}s...')
                    sleep_ms(backoff)
                    continue
                print(f'  Error: {err}')
                errors += len(batch)
                break

        write_json(PROGRESS_FILE, end)
        sleep_ms(BATCH_DELAY)

    print(f'\nDone.')
    print(f'  Created:    {created}')
    print(f'  Duplicates: {duplicates}')
    print(f'  Errors:     {errors}')
    print(f'\nNext: In HubSpot UI, filter contacts by contact_source = "linkedin-import" and set them as non-marketing.')


if __name__ == '__main__':
    main()
